File: solutions/0938. Range Sum of BST/0938.cpp
// Time complexity: O(n)
// Space complexity: O(h)
class Solution {
 public:
  int rangeSumBST(TreeNode* root, int low, int high) {
    if (root == nullptr) {
      return 0;
    }
    if (root->val < low) {
      return rangeSumBST(root->right, low, high);
    }
    if (root->val > high) {
      return rangeSumBST(root->left, low, high);
    }

    return root->val + rangeSumBST(root->right, low, high) + 
            rangeSumBST(root->left, low, high);
  }
};

File: solutions/0938. Range Sum of BST/0938.rs
// Time complexity: O(n)
// Space complexity: O(h)
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn range_sum_bst(root: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> i32 {
        if let Some(node) = root {
            let node = node.borrow();
            if node.val < low {
                return Self::range_sum_bst(node.right.clone(), low, high);
            } else if node.val > high {
                return Self::range_sum_bst(node.left.clone(), low, high);
            } else {
                return node.val
                    + Self::range_sum_bst(node.right.clone(), low, high)
                    + Self::range_sum_bst(node.left.clone(), low, high);
            }
        }

        0
    }
}
